[Cocoa] Implement Alternate WebM Player Logic

youssefsoli opened the pull request:
#### 84d162afe41d5510f7b039ae9116c519db5df5cd
<pre>
[Cocoa] Implement Alternate WebM Player Logic
<a href="https://bugs.webkit.org/show_bug.cgi?id=241949">https://bugs.webkit.org/show_bug.cgi?id=241949</a>

Reviewed by Jean-Yves Avenard.

Added the logic for the alternate WebM media player, this includes audio,
video, multi-track, fullscreen, and picture-in-picture support.

* Source/WTF/wtf/PlatformEnableCocoa.h:
* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/platform/graphics/WebMResourceClient.cpp: Added.
(WebCore::WebMResourceClient::create):
(WebCore::WebMResourceClient::WebMResourceClient):
(WebCore::WebMResourceClient::stop):
(WebCore::WebMResourceClient::dataReceived):
(WebCore::WebMResourceClient::loadFailed):
(WebCore::WebMResourceClient::loadFinished):
* Source/WebCore/platform/graphics/cocoa/MediaPlayerPrivateWebM.h:
(WebCore::MediaPlayerPrivateWebM::timeFudgeFactor const):
* Source/WebCore/platform/graphics/cocoa/MediaPlayerPrivateWebM.mm:
(WebCore::MediaPlayerPrivateWebM::MediaPlayerPrivateWebM):
(WebCore::MediaPlayerPrivateWebM::~MediaPlayerPrivateWebM):
(WebCore::mimeTypeCache):
(WebCore::MediaPlayerPrivateWebM::supportsType):
(WebCore::MediaPlayerPrivateWebM::dataReceived):
(WebCore::MediaPlayerPrivateWebM::loadFailed):
(WebCore::MediaPlayerPrivateWebM::loadFinished):
(WebCore::MediaPlayerPrivateWebM::load):
(WebCore::MediaPlayerPrivateWebM::cancelLoad):
(WebCore::MediaPlayerPrivateWebM::platformLayer const):
(WebCore::MediaPlayerPrivateWebM::play):
(WebCore::MediaPlayerPrivateWebM::pause):
(WebCore::MediaPlayerPrivateWebM::paused const):
(WebCore::MediaPlayerPrivateWebM::setPageIsVisible):
(WebCore::MediaPlayerPrivateWebM::currentMediaTime const):
(WebCore::MediaPlayerPrivateWebM::seek):
(WebCore::MediaPlayerPrivateWebM::setRateDouble):
(WebCore::MediaPlayerPrivateWebM::effectiveRate const):
(WebCore::MediaPlayerPrivateWebM::setVolume):
(WebCore::MediaPlayerPrivateWebM::setMuted):
(WebCore::MediaPlayerPrivateWebM::seekable const):
(WebCore::MediaPlayerPrivateWebM::buffered const):
(WebCore::MediaPlayerPrivateWebM::setBufferedRanges):
(WebCore::MediaPlayerPrivateWebM::updateBufferedFromTrackBuffers):
(WebCore::MediaPlayerPrivateWebM::setNaturalSize):
(WebCore::MediaPlayerPrivateWebM::setHasAudio):
(WebCore::MediaPlayerPrivateWebM::setHasVideo):
(WebCore::MediaPlayerPrivateWebM::setDuration):
(WebCore::MediaPlayerPrivateWebM::setNetworkState):
(WebCore::MediaPlayerPrivateWebM::setReadyState):
(WebCore::MediaPlayerPrivateWebM::characteristicsChanged):
(WebCore::MediaPlayerPrivateWebM::createVideoFullscreenLayer):
(WebCore::MediaPlayerPrivateWebM::setVideoFullscreenLayer):
(WebCore::MediaPlayerPrivateWebM::setVideoFullscreenFrame):
(WebCore::MediaPlayerPrivateWebM::requiresTextTrackRepresentation const):
(WebCore::MediaPlayerPrivateWebM::syncTextTrackBounds):
(WebCore::MediaPlayerPrivateWebM::setTextTrackRepresentation):
(WebCore::MediaPlayerPrivateWebM::enqueueSample):
(WebCore::MediaPlayerPrivateWebM::reenqueSamples):
(WebCore::MediaPlayerPrivateWebM::reenqueueMediaForTime):
(WebCore::MediaPlayerPrivateWebM::notifyClientWhenReadyForMoreSamples):
(WebCore::MediaPlayerPrivateWebM::canSetMinimumUpcomingPresentationTime const):
(WebCore::MediaPlayerPrivateWebM::setMinimumUpcomingPresentationTime):
(WebCore::MediaPlayerPrivateWebM::clearMinimumUpcomingPresentationTime):
(WebCore::MediaPlayerPrivateWebM::isReadyForMoreSamples):
(WebCore::MediaPlayerPrivateWebM::didBecomeReadyForMoreSamples):
(WebCore::MediaPlayerPrivateWebM::provideMediaData):
(WebCore::MediaPlayerPrivateWebM::trackDidChangeSelected):
(WebCore::MediaPlayerPrivateWebM::trackDidChangeEnabled):
(WebCore::MediaPlayerPrivateWebM::didParseInitializationData):
(WebCore::MediaPlayerPrivateWebM::didEncounterErrorDuringParsing):
(WebCore::MediaPlayerPrivateWebM::didProvideMediaDataForTrackId):
(WebCore::MediaPlayerPrivateWebM::append):
(WebCore::MediaPlayerPrivateWebM::abort):
(WebCore::MediaPlayerPrivateWebM::flush):
(WebCore::MediaPlayerPrivateWebM::flushIfNeeded):
(WebCore::MediaPlayerPrivateWebM::flushTrack):
(WebCore::MediaPlayerPrivateWebM::flushVideo):
(WebCore::MediaPlayerPrivateWebM::flushAudio):
(WebCore::MediaPlayerPrivateWebM::addTrackBuffer):
(WebCore::MediaPlayerPrivateWebM::ensureLayer):
(WebCore::MediaPlayerPrivateWebM::addAudioRenderer):
(WebCore::MediaPlayerPrivateWebM::removeAudioRenderer):
(WebCore::MediaPlayerPrivateWebM::destroyLayer):
(WebCore::MediaPlayerPrivateWebM::destroyAudioRenderer):
(WebCore::MediaPlayerPrivateWebM::destroyAudioRenderers):
(WebCore::MediaPlayerPrivateWebM::clearTracks):
(WebCore::MediaPlayerPrivateWebM::registerMediaEngine):
(WebCore::MediaPlayerPrivateWebM::isAvailable):

Canonical link: <a href="https://commits.webkit.org/251913@main">https://commits.webkit.org/251913@main</a>
</pre>

youssefsoli commented on Source/WebCore/WebCore.xcodeproj/project.pbxproj:
```diff
@@ -14653,10 +14655,10 @@
 		ADDF1AD41257CC1A1133B644 /* RenderSVGPath.cpp */ = {isa = PBXFileReference; fileEncoding = 4; lastKnownFileType = sourcecode.cpp.cpp; path = RenderSVGPath.cpp; sourceTree = "<group>"; };
+		ADDF1AD41257CD9A0003A759 /* LegacyRenderSVGPath.cpp */ = {isa = PBXFileReference; fileEncoding = 4; lastKnownFileType = sourcecode.cpp.cpp; path = LegacyRenderSVGPath.cpp; sourceTree = "<group>"; };
 		ADDF1AD51257CC1A1133B644 /* RenderSVGPath.h */ = {isa = PBXFileReference; fileEncoding = 4; lastKnownFileType = sourcecode.c.h; path = RenderSVGPath.h; sourceTree = "<group>"; };
+		ADDF1AD51257CD9A0003A759 /* LegacyRenderSVGPath.h */ = {isa = PBXFileReference; fileEncoding = 4; lastKnownFileType = sourcecode.c.h; path = LegacyRenderSVGPath.h; sourceTree = "<group>"; };
```
Unrelated sorting changes

youssefsoli commented on Source/WTF/wtf/PlatformEnableCocoa.h:
```diff
@@ -51,7 +51,7 @@
 #endif
 
-#if !defined(ENABLE_ALTERNATE_WEBM_PLAYER) && (PLATFORM(MAC) || PLATFORM(IOS) || PLATFORM(MACCATALYST))
+#if !defined(ENABLE_ALTERNATE_WEBM_PLAYER) && !PLATFORM(MACCATALYST) && !PLATFORM(WATCHOS) && !PLATFORM(APPLETV)
```
This was changed to address `PLATFORM(MACCATALYST)` not having `MEDIA_SOURCE` enabled.

jyavenard commented on Source/WebCore/platform/graphics/WebMResourceClient.cpp:
```diff
@@ -0,0 +1,93 @@
+
+namespace WebCore {
+
+WeakPtr<WebMResourceClient> WebMResourceClient::create(WeakPtr<WebMResourceClientParent>&& parent, PlatformMediaResourceLoader& loader, ResourceRequest&& request)
```
It's very unusual to return a WeakPtr or to even take a WeakPtr

It should take a ref in the constructor, up to it to only keep a WeakPtr

jyavenard reviewed (commented):
LGTM other than nits above, need to fix compilation failure. Will r+ once it's green

jyavenard commented on Source/WebCore/platform/graphics/WebMResourceClient.cpp:
```diff
@@ -0,0 +1,93 @@
+    return result;
+}
+
+WebMResourceClient::WebMResourceClient(WeakPtr<WebMResourceClientParent>&& parent, Ref<PlatformMediaResource>&& resource)
```
This should take a `WebmResourceClientParent&`

eric-carlson commented on Source/WebCore/platform/graphics/cocoa/MediaPlayerPrivateWebM.h:
```diff
@@ -29,89 +29,220 @@
+    // PlatformMediaResourceClient
+    void dataReceived(const SharedBuffer&) final;
+    void loadFailed(const ResourceError&) final;
+    void loadFinished(const FragmentedSharedBuffer&) final;
```
Do these need to be public?

eric-carlson commented on Source/WebCore/platform/graphics/cocoa/MediaPlayerPrivateWebM.h:
```diff
@@ -29,89 +29,220 @@
     MediaTime currentMediaTime() const final;
     MediaTime durationMediaTime() const final { return m_duration; };
-    
+    MediaTime startTime() const final { return MediaTime::zeroTime(); };
+    MediaTime initialTime() const final { return MediaTime::zeroTime(); };
```
Will these always be zero for WebM files? If not, it would be better to move them to the .cpp file and call `notImplemented()`

eric-carlson commented on Source/WebCore/platform/graphics/cocoa/MediaPlayerPrivateWebM.h:
```diff
@@ -29,89 +29,220 @@
+    void didParseInitializationData(InitializationSegment&&);
+    void didEncounterErrorDuringParsing(int32_t);
+    void didProvideMediaDataForTrackId(Ref<MediaSampleAVFObjC>&&, uint64_t trackId, const String& mediaType);
+
```
This blank line isn't needed

eric-carlson commented on Source/WebCore/platform/graphics/cocoa/MediaPlayerPrivateWebM.h:
```diff
@@ -29,89 +29,220 @@
+
     FloatSize m_naturalSize;
     bool m_hasAudio { false };
     bool m_hasVideo { false };
```
If you group all of the `bool` member variables together at the end of the object, the compiler may be able to make the object slightly smaller

eric-carlson commented on Source/WebCore/platform/graphics/cocoa/MediaPlayerPrivateWebM.mm:
```diff
@@ -70,73 +113,956 @@
+    if (!m_hasVideo && !m_hasAudio) {
+        setNetworkState(MediaPlayer::NetworkState::FormatError);
+        setReadyState(MediaPlayer::ReadyState::HaveNothing);
+        return;
```
It is probably worth logging an error here

eric-carlson commented on Source/WebCore/platform/graphics/cocoa/MediaPlayerPrivateWebM.mm:
```diff
@@ -70,73 +113,956 @@
+
+    auto mimeType = m_player->contentMIMEType();
+    if (mimeType.isEmpty() || !mimeTypeCache().contains(mimeType)) {
+        setNetworkState(MediaPlayer::NetworkState::FormatError);
```
Ditto

eric-carlson commented on Source/WebCore/platform/graphics/cocoa/MediaPlayerPrivateWebM.mm:
```diff
@@ -70,73 +113,956 @@
+    m_resourceClient = WebMResourceClient::create(WeakPtr { *this }, *loader, WTFMove(request));
+    
+    if (!m_resourceClient) {
+        setNetworkState(MediaPlayer::NetworkState::NetworkError);
+        setReadyState(MediaPlayer::ReadyState::HaveNothing);
```
Ditto

eric-carlson commented on Source/WebCore/platform/graphics/cocoa/MediaPlayerPrivateWebM.mm:
```diff
@@ -70,73 +113,956 @@
+            return;
+
+        [m_displayLayer enqueueSampleBuffer:platformSample.sample.cmSampleBuffer];
+
```
if you add an early return here you don't need the `else` below

eric-carlson commented on Source/WebCore/platform/graphics/cocoa/MediaPlayerPrivateWebM.mm:
```diff
@@ -70,73 +113,956 @@
+
+    if (!selected && m_enabledVideoTrackID == trackId) {
+        m_enabledVideoTrackID = -1;
+    } else if (selected) {
```
Ditto

eric-carlson commented on Source/WebCore/platform/graphics/cocoa/MediaPlayerPrivateWebM.mm:
```diff
@@ -70,73 +113,956 @@
+                    weakThis->didBecomeReadyForMoreSamples(trackId);
+            }];
+        }
+    } else if (m_audioRenderers.contains(trackId)) {
```
Ditto

eric-carlson commented on Source/WebCore/platform/graphics/cocoa/MediaPlayerPrivateWebM.mm:
```diff
@@ -70,73 +113,956 @@
+
+    if (!enabled)
+        removeAudioRenderer(trackId);
+    else {
```
Ditto

eric-carlson commented on Source/WebCore/platform/graphics/cocoa/MediaPlayerPrivateWebM.mm:
```diff
@@ -70,73 +113,956 @@
+    if (!trackBuffer)
+        return;
+
+    Ref<MediaSampleAVFObjC> sample = WTFMove(originalSample);
```
`auto`

eric-carlson commented on Source/WebCore/platform/graphics/cocoa/MediaPlayerPrivateWebM.mm:
```diff
@@ -70,73 +113,956 @@
+    DEBUG_LOG(LOGIDENTIFIER, trackId);
+
+    if (trackId == m_enabledVideoTrackID)
+        flushVideo();
```
Early return

youssefsoli committed:
[Cocoa] Implement Alternate WebM Player Logic
https://bugs.webkit.org/show_bug.cgi?id=241949

Reviewed by Jean-Yves Avenard.

Added the logic for the alternate WebM media player, this includes audio,
video, multi-track, fullscreen, and picture-in-picture support.

* Source/WTF/wtf/PlatformEnableCocoa.h:
* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/platform/graphics/WebMResourceClient.cpp: Added.
(WebCore::WebMResourceClient::create):
(WebCore::WebMResourceClient::WebMResourceClient):
(WebCore::WebMResourceClient::stop):
(WebCore::WebMResourceClient::dataReceived):
(WebCore::WebMResourceClient::loadFailed):
(WebCore::WebMResourceClient::loadFinished):
* Source/WebCore/platform/graphics/cocoa/MediaPlayerPrivateWebM.h:
(WebCore::MediaPlayerPrivateWebM::timeFudgeFactor const):
* Source/WebCore/platform/graphics/cocoa/MediaPlayerPrivateWebM.mm:
(WebCore::MediaPlayerPrivateWebM::MediaPlayerPrivateWebM):
(WebCore::MediaPlayerPrivateWebM::~MediaPlayerPrivateWebM):
(WebCore::mimeTypeCache):
(WebCore::MediaPlayerPrivateWebM::supportsType):
(WebCore::MediaPlayerPrivateWebM::dataReceived):
(WebCore::MediaPlayerPrivateWebM::loadFailed):
(WebCore::MediaPlayerPrivateWebM::loadFinished):
(WebCore::MediaPlayerPrivateWebM::load):
(WebCore::MediaPlayerPrivateWebM::cancelLoad):
(WebCore::MediaPlayerPrivateWebM::platformLayer const):
(WebCore::MediaPlayerPrivateWebM::play):
(WebCore::MediaPlayerPrivateWebM::pause):
(WebCore::MediaPlayerPrivateWebM::paused const):
(WebCore::MediaPlayerPrivateWebM::setPageIsVisible):
(WebCore::MediaPlayerPrivateWebM::currentMediaTime const):
(WebCore::MediaPlayerPrivateWebM::seek):
(WebCore::MediaPlayerPrivateWebM::setRateDouble):
(WebCore::MediaPlayerPrivateWebM::effectiveRate const):
(WebCore::MediaPlayerPrivateWebM::setVolume):
(WebCore::MediaPlayerPrivateWebM::setMuted):
(WebCore::MediaPlayerPrivateWebM::seekable const):
(WebCore::MediaPlayerPrivateWebM::buffered const):
(WebCore::MediaPlayerPrivateWebM::setBufferedRanges):
(WebCore::MediaPlayerPrivateWebM::updateBufferedFromTrackBuffers):
(WebCore::MediaPlayerPrivateWebM::setNaturalSize):
(WebCore::MediaPlayerPrivateWebM::setHasAudio):
(WebCore::MediaPlayerPrivateWebM::setHasVideo):
(WebCore::MediaPlayerPrivateWebM::setDuration):
(WebCore::MediaPlayerPrivateWebM::setNetworkState):
(WebCore::MediaPlayerPrivateWebM::setReadyState):
(WebCore::MediaPlayerPrivateWebM::characteristicsChanged):
(WebCore::MediaPlayerPrivateWebM::createVideoFullscreenLayer):
(WebCore::MediaPlayerPrivateWebM::setVideoFullscreenLayer):
(WebCore::MediaPlayerPrivateWebM::setVideoFullscreenFrame):
(WebCore::MediaPlayerPrivateWebM::requiresTextTrackRepresentation const):
(WebCore::MediaPlayerPrivateWebM::syncTextTrackBounds):
(WebCore::MediaPlayerPrivateWebM::setTextTrackRepresentation):
(WebCore::MediaPlayerPrivateWebM::enqueueSample):
(WebCore::MediaPlayerPrivateWebM::reenqueSamples):
(WebCore::MediaPlayerPrivateWebM::reenqueueMediaForTime):
(WebCore::MediaPlayerPrivateWebM::notifyClientWhenReadyForMoreSamples):
(WebCore::MediaPlayerPrivateWebM::canSetMinimumUpcomingPresentationTime const):
(WebCore::MediaPlayerPrivateWebM::setMinimumUpcomingPresentationTime):
(WebCore::MediaPlayerPrivateWebM::clearMinimumUpcomingPresentationTime):
(WebCore::MediaPlayerPrivateWebM::isReadyForMoreSamples):
(WebCore::MediaPlayerPrivateWebM::didBecomeReadyForMoreSamples):
(WebCore::MediaPlayerPrivateWebM::provideMediaData):
(WebCore::MediaPlayerPrivateWebM::trackDidChangeSelected):
(WebCore::MediaPlayerPrivateWebM::trackDidChangeEnabled):
(WebCore::MediaPlayerPrivateWebM::didParseInitializationData):
(WebCore::MediaPlayerPrivateWebM::didEncounterErrorDuringParsing):
(WebCore::MediaPlayerPrivateWebM::didProvideMediaDataForTrackId):
(WebCore::MediaPlayerPrivateWebM::append):
(WebCore::MediaPlayerPrivateWebM::abort):
(WebCore::MediaPlayerPrivateWebM::flush):
(WebCore::MediaPlayerPrivateWebM::flushIfNeeded):
(WebCore::MediaPlayerPrivateWebM::flushTrack):
(WebCore::MediaPlayerPrivateWebM::flushVideo):
(WebCore::MediaPlayerPrivateWebM::flushAudio):
(WebCore::MediaPlayerPrivateWebM::addTrackBuffer):
(WebCore::MediaPlayerPrivateWebM::ensureLayer):
(WebCore::MediaPlayerPrivateWebM::addAudioRenderer):
(WebCore::MediaPlayerPrivateWebM::removeAudioRenderer):
(WebCore::MediaPlayerPrivateWebM::destroyLayer):
(WebCore::MediaPlayerPrivateWebM::destroyAudioRenderer):
(WebCore::MediaPlayerPrivateWebM::destroyAudioRenderers):
(WebCore::MediaPlayerPrivateWebM::clearTracks):
(WebCore::MediaPlayerPrivateWebM::registerMediaEngine):
(WebCore::MediaPlayerPrivateWebM::isAvailable):

Canonical link: https://commits.webkit.org/251913@main
